### PR TITLE
move from telescope/path to plenary/path

### LINKED ISSUE #387: Move from `telescope/path` to `plenary/path`
I think everything from `telescope.path` can already be found in `plenary.path` so we should use that module instead and remove `telescope.path`

### PR DESCRIPTION
**Pull request by anott03**
Closes #387

### CODE REVIEW

**Comment by anott03** [comment]
Some of the changes in the diffs don't make sense... it's showing a bunch of changes regarding the names actions that shouldn't be there. When comparing my code to what I see on master by hand it seems to match. Can someone take a look?

File: README.md
Comment (anott03) [inline README.md:281]: Example of what I mentioned in my comment. This isn't a change I'm making (as far as I can tell)... this is already what's on master.

**Comment by Conni2461** [comment]
You probably based this branch on your last branch rather than on master. You could try to rebase it on current master. Not sure if it will work tho 😆 

Note: I will look at your plenary work first. I probably won't look at this PR today :)

**Comment by anott03** [comment]
@Conni2461 I'll try rebasing - thanks :)

> Note: I will look at your plenary work first. I probably won't look at this PR today :)

Of course. We all have things we need to do :)

**Comment by anott03** [comment]
This needs to be worked on some more after changes being made to plenary

File: lua/telescope/actions/init.lua
Comment (Conni2461) [inline lua/telescope/actions/init.lua:123]: Can we check if we still need `vim.fn.fnameescape` when doing path:new? @elianiva 

Also I kinda wanna move all `vim.fn.getcwd()` to `vim.loop.cwd()` to reduce our use of vimL

**Comment by Conni2461** [comment]
@anott03 Could you check if this one will be fixed when using `path:new` https://github.com/nvim-telescope/telescope.nvim/issues/394

Also did you have any luck with rebasing?

File: lua/telescope/actions/init.lua
Comment (elianiva) [reply lua/telescope/actions/init.lua:123]: was about to try this but I got this error.
```
...packer/start/telescope.nvim/lua/telescope/make_entry.lua:122: attempt to call method 'make_relative' (a nil value)
```

Comment (Conni2461) [reply lua/telescope/actions/init.lua:123]: Oh yeah requires a plenary PR : https://github.com/nvim-lua/plenary.nvim/pull/60

I am not sure if its done already 😆 but thanks for testing

Comment (elianiva) [reply lua/telescope/actions/init.lua:123]: Ah, so that's that I was missing :laughing:
Unfortunately we still need to escape it. You think it's a good idea to escape it ourself instead of relying on `fnameescape`?
My issue isn't that common and the only thing that's causing it is because `$` conflict with bash/zsh variable naming.

Comment (Conni2461) [reply lua/telescope/actions/init.lua:123]: Could you try this path for plenary on top of that PR:
```diff
diff --git a/lua/plenary/path.lua b/lua/plenary/path.lua
index 33372f9..c42c97e 100644
--- a/lua/plenary/path.lua
+++ b/lua/plenary/path.lua
@@ -200,14 +200,15 @@ end
 
 function Path:make_relative(cwd)
   cwd = F.if_nil(cwd, self._cwd, cwd)
-  if self.filename:sub(1, #cwd) == cwd  then
+  local filename = self:absolute() or self.filename
+  if filename:sub(1, #cwd) == cwd  then
     local offset =  0
     -- if  cwd does ends in the os separator, we need to take it off
     if cwd:sub(#cwd, #cwd) ~= path.separator then
       offset = 1
     end
 
-    self.filename = self.filename:sub(#cwd + 1 + offset, #self.filename)
+    self.filename = filename:sub(#cwd + 1 + offset, #filename)
   end
 
   return self.filename
@@ -217,7 +218,8 @@ function Path:normalize(cwd)
   cwd = F.if_nil(cwd, self._cwd, cwd)
   self:make_relative(cwd)
   -- Substitute home directory w/ "~"
-  self.filename = self.filename:gsub("^" .. path.home, '~', 1)
+  local filename = self:absolute() or self.filename
+  self.filename = filename:gsub("^" .. path.home, '~', 1)
   -- Remove double path seps, it's annoying
   self.filename = self.filename:gsub(path.sep .. path.sep, path.sep)
   print(self.filename)
@@ -247,7 +249,7 @@ local shorten = (function()
 end)()
 
 function Path:shorten()
-  return shorten(self.filename)
+  return shorten(self:absolute() or self.filename)
 end
 
 function Path:mkdir(opts)
```

Comment (elianiva) [reply lua/telescope/actions/init.lua:123]: Ok it's sorta works. I can open the file but it's not a valid file I think. Its content is different from what I have and I can't the file.

Comment (Conni2461) [reply lua/telescope/actions/init.lua:123]: Hmm okay. I think `Path:new` should handle the escaping better than vimL 🤣 . I will take a look at it later today. Shouldn't be to hard to create a file with `$` and do some testing 😆 Or maybe @anott03 wants to look at it :)

Still thanks for testing :)

Comment (elianiva) [reply lua/telescope/actions/init.lua:123]: no problem :)

**Comment by anott03** [comment]
@Conni2461 sure I'll take a look at #394 today.

As for the rebasing, I took a step back from telescope and was trying to get everything done with plenary first (which I think I have, so I'll get back to this today or tomorrow).

**Comment by anott03** [comment]
@Conni2461 I just rebased against master, and it seems to have worked (all tests pass, it seems to be running fine, etc.). As for #394, I tested it in the following way:
1. Checkout telescope master
2. `cd ~/.config/nvim` (which is a symlink to `~/repos/dotfiles/nvim`
3. run `:Telescope find_files`
I then repeated the above steps on my branch, `plenaryPath`. The results I yielded were the same for both branches, and that is that it successfully listed the files I had in that directory. I'm not sure if that means the issue has been solved, or if I need to take other steps to reproduce it. I'll also perform these tests on a symlink that doesn't point to a git repo and see if that makes a difference.

**Comment by tjdevries** [comment]
(What do you need me to do :laughing: )

**Comment by anott03** [comment]
@tjdevries Unless there are changes that need to be made to this code before it can be merged (stuff to resolve the failed linting test maybe?) I'd say there isn't much to do regarding this issue until the PR to plenary has been merged since this is dependent on that. It's probably worth it to see if someone can recreate #394. I've tried and haven't been able to.

**Comment by anott03** [comment]
@Conni2461 I don't think we should wait on `Path:escape` before merging this... I think it makes more sense just to resolve the conflicts and merge this as is and then create a separate PR for escaping paths if necessary.

**Comment by Conni2461** [comment]
Works for me. I am currently really busy (sorry exams and stuff) so i won't be around that much but i 100% trust @elianiva if he wants to finish it with you :)

**Comment by anott03** [comment]
No worries - good luck with the exams! I'll talk to @elianiva here or on gitter when he's free :)

**Comment by elianiva** [comment]
probably better talk it here, Gitter is a bit janky for me. It deletes DM from other people randomly for some reason 😆

**Comment by tjdevries** [comment]
Do you need anything from me?

**Comment by anott03** [comment]
@tjdevries I believe this is good to go - it just needs to be reviewed by someone. I had talked with Conni about waiting until we have a `Path:escape` in plenary but I think we should just merge this as is and then open a new PR if/when an escape function comes around.

**Comment by anott03** [comment]
@elianiva Can you take a look at the second conflict in `init.lua`? That's the only one I'm not quite sure how to resolve yet...

**Comment by elianiva** [comment]
remove [this part](https://github.com/nvim-telescope/telescope.nvim/blob/e94fdff12359d9dc3f93830c92f15d9fc8121e11/lua/telescope/actions/init.lua#L112-L130) from your branch and edit [this file](https://github.com/nvim-telescope/telescope.nvim/blob/6e941e0ecec1ab6a1b4ce40c693ef1272c505abb/lua/telescope/actions/set.lua) accordingly, specifically [this line](https://github.com/nvim-telescope/telescope.nvim/blob/d4bf1181ea6ef9aec675ff404d65098cbb91d9b1/lua/telescope/actions/set.lua#L4) and [this line](https://github.com/nvim-telescope/telescope.nvim/blob/d4bf1181ea6ef9aec675ff404d65098cbb91d9b1/lua/telescope/actions/set.lua#L111)

 We've moved the action that was previously in `actions/init.lua` to `actions/set.lua`. Hope it helps :)

**Comment by anott03** [comment]
Makes sense - thanks :)

**Comment by anott03** [comment]
I ended up resetting and doing everything over again because it turned out to be easier than rebasing against the latest changes (at least for me). In the processing of doing everything I realized that not only did `telescope.path` have some path related functions, but so did `utils`, so this now gets rid of any path functions in `utils` in favor of plenary as well.

@Conni2461 and @elianiva can one of you check this out?

File: lua/telescope/builtin/internal.lua
Comment (Conni2461) [inline lua/telescope/builtin/internal.lua:309]: Can we do `Path:new({data_dir, options, options.lua}):absolute()` to get the filepath (or similar)

File: lua/telescope/make_entry.lua
Comment (Conni2461) [inline lua/telescope/make_entry.lua:101]: Same here we shouldn't do concats manually anymore

Comment (Conni2461) [reply lua/telescope/make_entry.lua:101]: Or wait
Maybe we should here
just for performance. This code happens a lot

File: lua/telescope/path.lua
Comment (Conni2461) [inline lua/telescope/path.lua:11]: maybe we should just do. Similar to actions deprecated message
```lua
path.__index = function()
  print('deprecated message')
end
```

Comment (Conni2461) [reply lua/telescope/path.lua:11]: https://github.com/nvim-telescope/telescope.nvim/blob/2c4efc2f8af0575526f7b5dcc35fba9e0f1cc944/lua/telescope/actions/init.lua#L22-L35

Comment (Conni2461) [reply lua/telescope/path.lua:11]: Just can just always do a error. You don't need to do the string.finds

File: lua/telescope/utils.lua
Comment (Conni2461) [inline lua/telescope/utils.lua:232]: We should not doing concats here

**Comment by elianiva** [comment]
is there any reason why they're called `path` *or* `pathlib` ? I personally prefer `pathlib` to avoid `path.path.sep`, but I saw on some places, it's just called as `path` and some places called it as `pathlib`.

**Comment by anott03** [comment]
I'm guessing that's just an artifact of different code being written at different times. I just left the name as whatever it was before, but I agree that `pathlib` is a bit nicer. I've also been considering have `plenary.path` return `sep` since `path.path.sep` seems redundant.

**Comment by anott03** [comment]
@Conni2461 I've finally gotten back to work on this! I reset against master and everything seems to work with `plenary.path` now.

File: lua/telescope/builtin/internal.lua
Comment (Conni2461) [inline lua/telescope/builtin/internal.lua:334]: I think i already commented this :rofl: but: can we do `Path:new({utils.data_directory(), 'options', 'options.lua'}):absolute()` or so

File: lua/telescope/actions/set.lua
Comment (Conni2461) [inline lua/telescope/actions/set.lua:16]: Maybe we should make it more consistant. Sometimes `path` is written starting with a big letter and sometimes not.

If i have to choose i would take `Path`

File: lua/telescope/previewers/buffer_previewer.lua
Comment (Conni2461) [inline lua/telescope/previewers/buffer_previewer.lua:81]: I actually prefer `_read_async` here i know its kinda cool that it detects it automatically but we can save the `if` check here

File: lua/telescope/utils.lua
Comment (Conni2461) [inline lua/telescope/utils.lua:3]: Yep def some consistency when it comes to naming

**Comment by Conni2461** [comment]
Left some nitpicks

File: lua/telescope/actions/set.lua
Comment (anott03) [reply lua/telescope/actions/set.lua:16]: I agree - I had just tried to leave things as they were, but `Path` seems better to me.

**Comment by anott03** [comment]
Given that path is now from an external library (plenary), maybe `pathlib` is the best name. Either that or `Path`...

**Comment by anott03** [comment]
@Conni2461 As I've been working on this, I realized that `find_files` and `git_files` are both broken. It seems more general though - any of the builtin pickers that sort files seem to be broken. If you have a chance, can you take a look?

**Comment by Conni2461** [comment]
I can look at it. But can you rebase or so. Somehow the history and the diff looks super odd. Its hard to go through the changes with that :rofl: Ping me if its done.

**Comment by l-kershaw** [comment]
@anott03 I think the problem that you are encountering is probably coming from the fact that Telescope's `make_relative` has different parameters to Plenary's `make_relative`.
The first has inputs `filepath` and `cwd` which are both strings. The second has inputs of `self` and `cwd` where `self` is a `Path` object and `cwd` is a string.

This ends up having plenary's make relative being given a string where it expects a `Path` object, so it passes `self.filename` (which is `nil`) to the `clean` function. This then causes errors when using `gsub` in the current way.

---

As mentioned by @Conni2461, it is a lot easier to help with bugs if you rebase your PR branch to the latest `master` rather than merge, as the latter causes there to be lots of irrelevant changes in the history of this branch (as they have already been merged into Telescope `master`).

**Comment by anott03** [comment]
@l-kershaw @Conni2461 apologies for disappearing for a while - I'm fairly busy for the next week but after that I'll be able to spend more time working on this. I've rebased against master, so the history should be cleared up now (and hopefully I didn't break anything in the process🤞).

File: lua/telescope/make_entry.lua
Comment (Conni2461) [inline lua/telescope/make_entry.lua:447]: this part looks wrong. `Path:new(bufname):normalize(cwd)`

File: lua/telescope/make_entry.lua
Comment (Conni2461) [inline lua/telescope/make_entry.lua:816]: probably also `Path:new(vim.fn.expand('%')):normalize(cwd)`

**Comment by Conni2461** [comment]
Also we wanna remove `telescope.path` so we should replace each function with a error message that says. "Please use plenary.path" or similar. So extensions know that they have to move to the new way and then remove the whole module in a month or so.

Similar to https://github.com/nvim-telescope/telescope.nvim/blob/49b86b4610199d879f539d9193fbee45792ba712/lua/telescope/utils.lua#L420-L434 for example (i think we could remove those functions now btw)

lets finish this pr in the next two days. I am here if you need me :)

**Comment by anott03** [comment]
@Conni2461 how do we typically handle deprecating things like `path.separator` which are variables rather than functions?

**Comment by anott03** [comment]
Could we do something like this?
```lua
local path = setmetatable({}, {
  __index = function()
    error("telescope.path is deprecated. please use plenary.path instead")
  end
})

return path

```

**Comment by Conni2461** [comment]
Yes thats better for our case because we deprecate a whole module, not just a couple of functions :+1:

**Review by Conni2461 — APPROVED** [review]
LGTM

I need to test it, so i will run it today and see if something doesnt work. If all works out i'll merge this evening

Thank you very much :) and thank you for your patience :)

File: lua/telescope/builtin/files.lua
Comment (Conni2461) [inline lua/telescope/builtin/files.lua:254]: Hmm `Telescope file_browser` is broken because dirs no longer have the os sep at the end of the file. We need to make sure that it stays there.

I dont even this this change is coursing it but changes in `gen_from_file`. Maybe we should write a completly new `entry_maker` for file_browser. What do you think?

File: lua/telescope/builtin/files.lua
Comment (Conni2461) [inline lua/telescope/builtin/files.lua:287]: Pretty sure this is wrong. Because we are doing `Path:new(fpath)` later. I haven't tested this but maybe `fpath:` stuff that happens below might just work.

File: lua/telescope/builtin/files.lua
Comment (anott03) [reply lua/telescope/builtin/files.lua:254]: Yeah it's not 100% clear to me what you mean, but I'd be happy to take a crack at it later today :)

File: lua/telescope/builtin/files.lua
Comment (Conni2461) [reply lua/telescope/builtin/files.lua:287]: Right might be but we also do a string sub in the if. I keep it as string for now :)

File: lua/telescope/builtin/files.lua
Comment (Conni2461) [reply lua/telescope/builtin/files.lua]: Sorry i just did it myself :)

**Comment by Conni2461** [comment]
I talked with tj for a second. He wants to keep the module for now but throw an error. Just so extensions have a week or so to switch over to plenary.path

I am going to merge this now.

Thanks for all your work and your patience with me :)

**Comment by clason** [comment]
Shouldn't this get an entry in https://github.com/nvim-telescope/telescope.nvim/blob/master/doc/telescope_changelog.txt?

**Comment by Conni2461** [comment]
I forgot about this. Calling the function will yell at you tho :laughing:

**Comment by clason** [comment]
Yeah, but some explanation -- and migration guide! -- would be good since it's not as simple as replacing `require'telescope.path'` with `require'plenary.path'` ;)

**Comment by Conni2461** [comment]
I'll add one i noticed something else super small too too, i shouldn't have merged. Honest mistake. I was just glad right now that it was done and it didn't break anything inside telescope :laughing:

**Comment by anott03** [comment]
@Conni2461 thanks for all your help! Lmk if there's anything else you'd like me to do :)

**Pull request merged**
